### ISSUE TITLE
v0.52: fp_compute{Covariances,Correlations}

### ISSUE THREAD

**Issue by beanSnippet** [libqueso/queso#388]
If either ```fp_computeCovariances``` or ```fp_computeCorrelations``` is turned on in the input file, then I get this error:
```
Assertion `pqCorrMatrix(i,j) >= -1. - 1.e-8' failed.
pqCorrMatrix(i,j) = -nan
-1. - 1.e-8 = -1
computed correlation is out of range
basic/src/VectorSequence.C, line 2878
```
I get the same chains (for IP and FP) when I run with these options turned off.

**Comment by dmcdougall** [comment]
The offending commit is 7725a3b5de362f5bf88bed0298628a26e41da6c3.

This looks like we missed some subtle logic issues when refactoring our assert macros.

**Comment by dmcdougall** [comment]
Or perhaps not, since the correlation matrix has a nan element.

**Comment by dmcdougall** [comment]
Ok, so the reason you now get an error is because @roystgnr's macro updates fixed a bug.  If you look at [this](https://github.com/libqueso/queso/commit/7725a3b5de362f5bf88bed0298628a26e41da6c3#diff-7a01fb53baade23ab8784706267b9864L2728) line you'll see that a `<` comparison is executed.  When a comparison is done on a `NaN`, the result is `false`.  Given the test is computed whether or not a value *is* between certain bounds, a result of `false` is interpreted as "`NaN` is between the acceptable bounds", which is the incorrect interpretation.

The new macros, however, reverse the test.  So now the test reads "reject everything not between these bounds."  Remember for `NaN`s, the comparison test will result in `false`, which is now the correct interpretation.

What's my point?  The new macros are superior to the old macros.  So this means some investigation needs to be done regarding where and why this function is running into a `NaN`.

**Comment by dmcdougall** [comment]
And the reason you're getting a `NaN` here is because, from the example you emailed me, queso is trying to compute the correlations between things using a single posterior sample ([there is no check on the third argument to this function being 1](https://github.com/libqueso/queso/blob/dev/src/stats/src/StatisticalForwardProblem.C#L288)).  You need at least 2 samples to compute a variance.

Setting either `ip_mh_filteredChain_lag = 2` to get five samples, or `ip_mh_filteredChain_generate = 0` to use all ten samples fixes the problem.

That said, we can approach this two ways.  I can add a check to queso so that it either a) fails a little more gracefully when trying to compute the correlation with one sample; or b) add a check to skip the calculation (and not fail at all).

Does anyone have a preference regarding this?

**Comment by pbauman** [comment]
IMHO:

>  a) fails a little more gracefully when trying to compute the correlation with one sample

**Comment by dmcdougall** [comment]
It seems I spoke too soon.  Even with 10 samples I'm getting a `NaN`.

**Comment by dmcdougall** [comment]
Ok, I think we're finally there.

In the example you emailed me, your qoi values are all zero:

```
for (unsigned int i = 0; i < 1; i++) qoiValues[i] = i;
```

I changed this line to

```
for(unsigned int i = 0; i < 1; i++) qoiValues[i] = paramValues[i];
```

and I no longer run into a `NaN`.  The reason all zeroes results in a `NaN` is because QUESO is computing a variance, which ends up being zero.  This variance is the denominator for a correlation matrix and you can see that [here](https://github.com/libqueso/queso/blob/dev/src/basic/src/VectorSequence.C#L2862).  The `unifiedSampleVarianceQ` variable is zero in your case, and `gdb` confirms this:

```
(gdb) p unifiedSampleVarianceQ.operator[](0)
$1 = (double &) @0x626b00: 0
(gdb) p unifiedSampleVarianceP.operator[](0)
$2 = (double &) @0x61c300: 0.036032882342928646
```

In addition to the check that the number of samples is at least two, I can also add a check that the variance is positive for both parameters and quantities of interest and fail gracefully in this case.

**Comment by beanSnippet** [comment]
Sounds good. I know at some point I was getting a NaN with more than one sample and non-zero qoi values (outside the test function above), but that was a long time ago. I'll let you know what happens if I use this again.

**Comment by dmcdougall** [comment]
Addressed in #394.  Closing.
